lib/remap-functions/src/length.rs

```diff
@@ -0,0 +1,133 @@
+use remap::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct Length;
+
+impl Function for Length {
+    fn identifier(&self) -> &'static str {
+        "length"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            accepts: |v| matches!(v, Value::Array(_) | Value::Map(_)),
+            required: true,
+        }]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Result<Box<dyn Expression>> {
+        let value = arguments.required("value")?.boxed();
+
+        Ok(Box::new(LengthFn { value }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct LengthFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for LengthFn {
+    fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        use Value::*;
+
+        let value = self.value.execute(state, object)?;
+
+        match value {
+            Array(v) => Ok(Value::from(v.len() as i64)),
+            Map(v) => Ok(Value::from(v.len() as i64)),
+            Bytes(v) => Ok(Value::from(v.len() as i64)),
+            _ => Err("unsupported type".into()),
+        }
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        use value::Kind;
+
+        let value_def = self
+            .value
+            .type_def(state)
+            .fallible_unless(Kind::Bytes | Kind::Array | Kind::Map)
+            .with_constraint(Kind::Integer);
+
+        if value_def.kind.is_scalar() {
+            value_def.with_inner_type(None)
+        } else {
+            value_def
+        }
+    }
```

Comment:
My proposal would be to move this into the `with_constraint` function, since there's never a case where you want to keep the inner type def if the kind is known to only resolve to a scalar value.

Reply:
I agree, it'd be nice to move that there. We can drop the `with_inner_type` from the `includes()` remap function as part of this implementation and just remove `with_inner_type` altogether.

Reply:
I'm a little confused about what it would mean to move this "into" the `with_constraint` function given that the setters here form a sort of builder pattern

Reply:
Yeah, my interpretation, and opinion, is to move the actual logic into the `with_constraint` method so it is consistently applied for all scalars.

If you do that, I think you can drop the current implementation of `with_inner_type` as it was only added to do basically what you are doing here in the `includes()` remap function.

Reply:
@jszwedko Ah, okay, that clarifies it 100%.

Reply:
Okay, updated 👍🏼